lib/req/steps.ex

```diff
@@ -829,4 +825,31 @@ defmodule Req.Steps do
   defp parse_netrc([], nil, acc) do
     acc
   end
+
+  defp put_redirect_location(request, location) do
+    if String.starts_with?(location, "/") do
+      url = URI.parse(location)
+      update_in(request.url, &%{&1 | path: url.path, query: url.query})
+    else
+      url = URI.parse(location)
+      put_in(request.url, url)
+    end
+  end
+
+  defp remove_credentials_if_untrusted(%{location_trusted: true} = request, _), do: request
+
+  defp remove_credentials_if_untrusted(%{url: url} = request, location) do
+    if URI.parse(location).host == url.host do
```

Comment:
it seems we're calling `URI.parse(location)` here for the 2nd time, first time was in `put_redirect_location`. Could you change it to do the parsing just once? Maybe like this?

```elixir
  location = URI.parse(location)
  request =
    request
    |> remove_credentials_if_untrusted(location)
    |> put_redirect_location(location)
```

Reply:
To do this we have to replace the check of the first character in the location string in `put_redirect_location`. I would propose that instead of checking `String.starts_with?(location, "/")` we check `location_url.host` (where `location_url` is the parsed `location`), since the `:host` key of the `location_url` struct would be nil if the location is a relative path.

Alternatively we could pass both the string and the struct to `put_redirect_location` but that seems more messy.

Reply:
> we check `location_url.host`

sounds great!